Derive pci device_spec edpm_derive_pci_device_spec

vcandapp opened the pull request:
This role will generate the pci device_spec for sriov nic partitioning deployment.
The generated config file should be used to create device_spec configmap for custom nova dataplane service.

This has been re-factored from [PR](https://github.com/openstack-k8s-operators/edpm-ansible/pull/572)

Jaganathancse commented on roles/edpm_derive_pci_device_spec/tasks/main.yml:
```diff
@@ -0,0 +1,18 @@
+# under the License.
+
+- name: Derive the SR-IOV device_spec config map for nic-partitionin
+  ansible.builtin.import_tasks: derive_pci.yml
```
@vcandapp please change derive_pci.yml file name like derive_pci_device_spec.yml.

vcandapp replied:
Changed file name for more clarity

Jaganathancse commented on roles/edpm_derive_pci_device_spec/tasks/derive_pci.yml:
```diff
@@ -0,0 +1,48 @@
+  when:
+    - not sriov_device_spec_conf_check.stat.exists
+  become: true
+  block:
```
@vcandapp block is not needed.

Jaganathancse commented on roles/edpm_derive_pci_device_spec/tasks/derive_pci.yml:
```diff
@@ -0,0 +1,48 @@
+        sriov_device_spec_map: "{{ edpm_derive_sriov_device_spec_list }}"
+        sriov_phydev_map: "{{edpm_derive_pci_device_spec_physmap_file }}"
+        sriov_device_spec_out_file: "{{ edpm_derive_pci_device_spec_conf_dir }}/{{ edpm_derive_pci_device_spec_conf_file }}"
+      register: derive_pci_devicespec
```
@vcandapp line 47 and 48 are not required.

Jaganathancse replied:
derive_pci_devicespec.rc will throw error.

vcandapp replied:
This will be needed

Jaganathancse commented on roles/edpm_derive_pci_device_spec/tasks/derive_pci.yml:
```diff
@@ -0,0 +1,48 @@
+- name: Check valid input for edpm_derive_sriov_device_spec_list
+  ansible.builtin.fail:
+    msg: "List of sriov device_spec cannot be empty - edpm_derive_sriov_device_spec_list"
+  when: edpm_derive_sriov_device_spec_list | length > 0
```
@vcandapp Need to update condition length==0 for fail message.

vcandapp replied:
Done

Jaganathancse commented on docs/source/roles/role-edpm_derive_pci_device_spec.rst:
```diff
@@ -0,0 +1,26 @@
+
+.. code-block:: YAML
+
+    - name: "Configure DDP Package"
```
@vcandapp typo DDP

vcandapp replied:
Done

Jaganathancse commented on plugins/modules/edpm_derive_pci_passthrough_devicespec.py:
```diff
@@ -0,0 +1,650 @@
+  type: int
+'''
+
+_PASSTHROUGH_DEVSPEC_KEY = 'device_spec'
```
@vcandapp Please remove starting with '_'.  to maintain code similar like other module.

Jaganathancse replied:
for line 81 to 83.

Jaganathancse commented on plugins/modules/edpm_derive_pci_passthrough_devicespec.py:
```diff
@@ -0,0 +1,650 @@
+    pass
+
+
+def get_pci_field_val(
```
@vcandapp can we add description for all the functions?

Jaganathancse commented on plugins/modules/edpm_derive_pci_passthrough_devicespec.py:
```diff
@@ -0,0 +1,650 @@
+
+
+def get_pci_regex_pattern(config_regex: str, size: int, maxval: int, hex_value: str):
+    if config_regex in [ANY, REGEX_ANY]:
```
@vcandapp same description and also either we can specify type for argument variables in all the functions or remove it. looks like not consistent.

Jaganathancse commented on plugins/modules/edpm_derive_pci_passthrough_devicespec.py:
```diff
@@ -0,0 +1,650 @@
+
+
+def get_user_regex_from_addrdict(addr_dict):
+    if isinstance(addr_dict, dict):
```
same desciption missing.

Jaganathancse commented on plugins/modules/edpm_derive_pci_passthrough_devicespec.py:
```diff
@@ -0,0 +1,650 @@
+
+
+def get_allocated_pci_addresses(configs):
+    alloc_pci_info = []
```
same description missing

vcandapp replied:
Done

Jaganathancse commented on plugins/modules/edpm_derive_pci_passthrough_devicespec.py:
```diff
@@ -0,0 +1,650 @@
+
+
+def get_pci_passthrough_devicespec(user_config, pf, pci_addresses):
+    pci_passthrough_list = []
```
same description missing

Jaganathancse commented on plugins/modules/edpm_derive_pci_passthrough_devicespec.py:
```diff
@@ -0,0 +1,650 @@
+
+def get_passthrough_config_for_all_pf(user_config,
+                                      system_configs,
+                                      allocated_pci):
```
same description missing

Jaganathancse commented on plugins/modules/edpm_derive_pci_passthrough_devicespec.py:
```diff
@@ -0,0 +1,650 @@
+    configs = []
+    src_config = ""
+    # Currently only "device_spec" is valid
+    valid_spec = ('device_spec')
```
@vcandapp  can we reuse this variable PASSTHROUGH_DEVSPEC_KEY instead of valid_spec new variable again?

Jaganathancse commented on plugins/modules/edpm_derive_pci_passthrough_devicespec.py:
```diff
@@ -0,0 +1,650 @@
+                    sriov_phydev_map=None):
+    pci_passthrough = {}
+    system_configs = get_sriov_configs()
+    user_configs = sriov_device_spec
```
@vcandapp remove this line and specify user_configs as _run_derive_pci function argument sriov_device_spec,?
since no user_configs updates done. only for reading purpose.

vcandapp replied:
Done

Jaganathancse commented on plugins/modules/edpm_derive_pci_passthrough_devicespec.py:
```diff
@@ -0,0 +1,650 @@
+      - List of SR-IOV device_spec
+    type: list
+    required: true
+    default: ''
```
@vcandapp  used required but default is empty. mutually exclusive and get internal error

vcandapp replied:
Done

Jaganathancse commented on plugins/modules/edpm_derive_pci_passthrough_devicespec.py:
```diff
@@ -0,0 +1,650 @@
+      - ConfigMap file to be used in nic-partitioned deployment
+    type: str
+    required: true
+    default: '/var/lib/config-data/ansible-generated/derive_devicespec/20-sriov-nova.conf'
```
here too.

vcandapp replied:
Updated

vcandapp commented on plugins/modules/edpm_derive_pci_passthrough_devicespec.py:
```diff
@@ -0,0 +1,650 @@
+    type: str
+    required: false
+    default: ''
+  sriov_nova_conf_file:
```
Updated

Jaganathancse commented on roles/edpm_derive_pci_device_spec/tasks/derive_pci_device_spec.yml:
```diff
@@ -0,0 +1,51 @@
+      changed_when: derive_pci_devicespec.rc == 0
+    - name: Print result of derived device_spec 
+      debug:
+        msg: "{{ derive_pci_devicespec }}"
```
@vcandapp  Please use derive_pci_devicespec.device_spec instead of complete details

vcandapp replied:
Updated debug msg

Jaganathancse commented:
/lgtm

Jaganathancse reviewed (approved):
Looks good.

Jaganathancse commented:
/approved

Jaganathancse commented:
/approve

vcandapp commented:
**Did E-2-E testing with this PR in nic partitioned deployement, and here are the verified logs:**

[WARNING]: running playbook inside collection osp.edpm

PLAY [Derive sriov device_spec] ************************************************

TASK [Gathering Facts] *********************************************************
Thursday 14 March 2024  16:49:17 +0000 (0:00:00.035)       0:00:00.035 ******** 
ok: [tigon30]
ok: [tigon29]

TASK [osp.edpm.edpm_derive_pci_device_spec : Validating arguments against arg spec 'main' - The main entry point for the edpm_derive_pci_device_spec role.] ***
Thursday 14 March 2024  16:49:19 +0000 (0:00:01.507)       0:00:01.542 ******** 
ok: [tigon29] => {"changed": false, "msg": "The arg spec validation passed", "validate_args_context": {"argument_spec_name": "main", "name": "edpm_derive_pci_device_spec", "path": "/usr/share/ansible/collections/ansible_collections/osp/edpm/roles/edpm_derive_pci_device_spec", "type": "role"}}
ok: [tigon30] => {"changed": false, "msg": "The arg spec validation passed", "validate_args_context": {"argument_spec_name": "main", "name": "edpm_derive_pci_device_spec", "path": "/usr/share/ansible/collections/ansible_collections/osp/edpm/roles/edpm_derive_pci_device_spec", "type": "role"}}

TASK [osp.edpm.edpm_derive_pci_device_spec : Check valid input for edpm_derive_sriov_device_spec_list] ***
Thursday 14 March 2024  16:49:19 +0000 (0:00:00.064)       0:00:01.606 ******** 
skipping: [tigon29] => {"changed": false, "skip_reason": "Conditional result was False"}
skipping: [tigon30] => {"changed": false, "skip_reason": "Conditional result was False"}

TASK [osp.edpm.edpm_derive_pci_device_spec : Create neutron-sriov-agent directories] ***
Thursday 14 March 2024  16:49:19 +0000 (0:00:00.052)       0:00:01.659 ******** 
changed: [tigon29] => {"changed": true, "gid": 1000, "group": "cloud-admin", "mode": "0755", "owner": "cloud-admin", "path": "/var/lib/config-data/ansible-generated/derive_devicespec", "secontext": "unconfined_u:object_r:container_file_t:s0", "size": 6, "state": "directory", "uid": 1000}
changed: [tigon30] => {"changed": true, "gid": 1000, "group": "cloud-admin", "mode": "0755", "owner": "cloud-admin", "path": "/var/lib/config-data/ansible-generated/derive_devicespec", "secontext": "unconfined_u:object_r:container_file_t:s0", "size": 6, "state": "directory", "uid": 1000}

TASK [osp.edpm.edpm_derive_pci_device_spec : Check if edpm node has the sriov nova conf] ***
Thursday 14 March 2024  16:49:19 +0000 (0:00:00.403)       0:00:02.063 ******** 
ok: [tigon29] => {"changed": false, "stat": {"exists": false}}
ok: [tigon30] => {"changed": false, "stat": {"exists": false}}

TASK [osp.edpm.edpm_derive_pci_device_spec : Run derive_pci_passthrough_devicespec] ***
Thursday 14 March 2024  16:49:20 +0000 (0:00:00.379)       0:00:02.442 ******** 
changed: [tigon30] => {"changed": true, "device_spec": {"device_spec": [{"address": "0000:17:04.6", "trusted": "true"}, {"address": "0000:17:04.4", "trusted": "true"}, {"address": "0000:17:04.2", "trusted": "true"}, {"address": "0000:17:04.3", "trusted": "true"}]}, "message": "PCI device_spec is generated", "rc": 0, "sucess": false}
changed: [tigon29] => {"changed": true, "device_spec": {"device_spec": [{"address": "0000:17:04.6", "trusted": "true"}, {"address": "0000:17:04.4", "trusted": "true"}, {"address": "0000:17:04.2", "trusted": "true"}, {"address": "0000:17:04.3", "trusted": "true"}]}, "message": "PCI device_spec is generated", "rc": 0, "sucess": false}

TASK [osp.edpm.edpm_derive_pci_device_spec : Print result of derived device_spec] ***
Thursday 14 March 2024  16:49:20 +0000 (0:00:00.394)       0:00:02.836 ******** 
ok: [tigon29] => {
    "msg": {
        "device_spec": [
            {
                "address": "0000:17:04.6",
                "trusted": "true"
            },
            {
                "address": "0000:17:04.4",
                "trusted": "true"
            },
            {
                "address": "0000:17:04.2",
                "trusted": "true"
            },
            {
                "address": "0000:17:04.3",
                "trusted": "true"
            }
        ]
    }
}
ok: [tigon30] => {
    "msg": {
        "device_spec": [
            {
                "address": "0000:17:04.6",
                "trusted": "true"
            },
            {
                "address": "0000:17:04.4",
                "trusted": "true"
            },
            {
                "address": "0000:17:04.2",
                "trusted": "true"
            },
            {
                "address": "0000:17:04.3",
                "trusted": "true"
            }
        ]
    }
}

PLAY RECAP *********************************************************************
tigon29                    : ok=6    changed=2    unreachable=0    failed=0    skipped=1    rescued=0    ignored=0   
tigon30                    : ok=6    changed=2    unreachable=0    failed=0    skipped=1    rescued=0    ignored=0   
Thursday 14 March 2024  16:49:20 +0000 (0:00:00.146)       0:00:02.983 ******** 
=============================================================================== 
Gathering Facts --------------------------------------------------------- 1.51s
osp.edpm.edpm_derive_pci_device_spec : Create neutron-sriov-agent directories --- 0.40s
osp.edpm.edpm_derive_pci_device_spec : Run derive_pci_passthrough_devicespec --- 0.39s
osp.edpm.edpm_derive_pci_device_spec : Check if edpm node has the sriov nova conf --- 0.38s
osp.edpm.edpm_derive_pci_device_spec : Print result of derived device_spec --- 0.15s
osp.edpm.edpm_derive_pci_device_spec : Validating arguments against arg spec 'main' - The main entry point for the edpm_derive_pci_device_spec role. --- 0.06s
osp.edpm.edpm_derive_pci_device_spec : Check valid input for edpm_derive_sriov_device_spec_list --- 0.05s

Jaganathancse commented:
/approve

bshephar reviewed (approved):
Looks good. Glad to see some good test coverage and molecule tests. Be nice to have some end-to-end SR-IOV testing downstream to make sure we aren't missing anything.
